### PR TITLE
many: make kernel-modules components available on first boot

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Make sure that module shipped in kernel-modules components are available on first boot after auto-install. This does not consider yet installation using the snapd API.

### CODE REVIEW

File: gadget/install/kernel.go [hunk 1]
```diff
@@ -31,8 +31,21 @@ type KernelSnapInfo struct {
 	// MountPoint is the root of the files from the kernel snap
 	MountPoint string
 	// NeedsDriversTree will be set if a drivers tree needs to be
-	// build on installation
+	// built on installation
 	NeedsDriversTree bool
 	// IsCore is set if this is UC
 	IsCore bool
+	// modulesComps has the information for installed
```
Comment (andrewphelpsj) [inline gadget/install/kernel.go:38]:
```suggestion
	// ModulesComps has the information for installed
```

**Review by andrewphelpsj — APPROVED** [review]
Thanks, this looks good to me. Just some minor things.

File: overlord/devicestate/handlers_install.go [hunk 2]
```diff
@@ -274,11 +300,36 @@ func (m *DeviceManager) doSetupRunSystem(t *state.Task, _ *tomb.Tomb) error {
 	var installedSystem *install.InstalledSystemSideData
 	// run the create partition code
 	logger.Noticef("create and deploy partitions")
+
+	// Find out installed components in the temporary system, which will be
+	// the same as those snapd will install in the final system.
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:305]: is this true though?  components can have modes too in the model

**Review by pedronis — COMMENTED** [review]
question about one of the assumptions, also don't we need a spread test for this?

File: overlord/devicestate/handlers_install.go [hunk 3]
```diff
@@ -205,6 +205,32 @@ func (m *DeviceManager) doSetupUbuntuSave(t *state.Task, _ *tomb.Tomb) error {
...
+func kernelModulesComps(st *state.State, kernelName string) ([]*snap.ComponentInfo, error) {
+	var stateMap map[string]*snapstate.SnapState
+	if err := st.Get("snaps", &stateMap); err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	kernel, ok := stateMap[kernelName]
+	if !ok {
+		return nil, fmt.Errorf("internal error: %s is not installed", kernelName)
+	}
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/handlers_install.go:217]: yes, this should likely be snapstate.Get

File: boot/makebootable.go [hunk 4]
```diff
@@ -372,19 +379,27 @@ func makeRunnableSystem(model *asserts.Model, bootWith *BootableSet, observer Tr
...
-		if err := copyBootSnap(origDest.orig, origDest.destInfo, snapBlobDir); err != nil {
+		{orig: bootWith.BasePath, fileName: bootWith.Base.Filename()},
+		{orig: bootWith.KernelPath, fileName: bootWith.Kernel.Filename()},
+		{orig: bootWith.GadgetPath, fileName: bootWith.Gadget.Filename()}} {
+		if err := copyBootSnap(origDest.orig, origDest.fileName, snapBlobDir); err != nil {
+			return err
+		}
+	}
+	for _, kmod := range bootWith.KernelMods {
+		if err := copyBootSnap(kmod.CompPath, kmod.CompPlaceInfo.Filename(), snapBlobDir); err != nil {
```
Comment (bboozzoo) [inline boot/makebootable.go:402]: would this work?
```suggestion
		if err := copyBootSnap(kmod.CompPlaceInfo.MountFile(), kmod.CompPlaceInfo.Filename(), snapBlobDir); err != nil {
```

**Comment by alfonsosanchezbeato** [comment]
I have pushed changes to this now that look at the seed for normal installation, snapd API installation, and for factory reset. spread tests are missing still.

Depends on https://github.com/canonical/snapd/pull/14770

File: boot/makebootable.go [hunk 5]
```diff
@@ -56,6 +56,16 @@ type BootableSet struct {
...
+
+	// KernelMods contains kernel-modules components in the system.
+	KernelMods []BootableKModsComponents
+}
+
+// BootableComponent represents kernel-modules components, which are
+// needed as part of a BootableSet.
+type BootableKModsComponents struct {
+	CompPlaceInfo snap.ContainerPlaceInfo
+	CompPath      string
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply boot/makebootable.go]: CompPath is here the path in the seed, not the final place where the file is going to be placed in the system.

File: boot/makebootable_test.go [hunk 6]
```diff
@@ -497,7 +497,7 @@ func (s *makeBootable20Suite) TestMakeSystemRunnable16Fails(c *C) {
 	c.Assert(err, ErrorMatches, `internal error: cannot make pre-UC20 system runnable`)
 }
 
-func (s *makeBootable20Suite) testMakeSystemRunnable20(c *C, standalone, factoryReset, classic bool, fromInitrd bool) {
+func (s *makeBootable20Suite) testMakeSystemRunnable20(c *C, standalone, factoryReset, classic, fromInitrd, withKComps bool) {
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply boot/makebootable_test.go]: Changed

File: gadget/install/kernel.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply gadget/install/kernel.go]: Changed

File: overlord/devicestate/handlers_install.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/handlers_install.go]: Now this uses the seed for the different install cases

File: boot/makebootable.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply boot/makebootable.go:404]: CompPath is here the path in the seed, not the final place where the file is going to be placed in the system.

**Comment by alfonsosanchezbeato** [comment]
Thanks for the reviews, this is ready for another round now that https://github.com/canonical/snapd/pull/14770 is merged, although I plan to push also spread tests.

File: overlord/devicestate/handlers_install.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/devicestate/handlers_install.go]: Makes sense, this is outdated now though

File: overlord/devicestate/devicemgr.go [hunk 7]
```diff
@@ -2242,16 +2248,20 @@ func (m *DeviceManager) defaultRecoverySystem() (*DefaultRecoverySystem, error)
...
 // loadSystemAndEssentialSnaps loads information for the given label, which
-// includes system, gadget information, gadget and kernel snaps info,
-// and gadget and kernel seed snap info.
+// includes system, gadget information, gadget and kernel snaps info, and
+// gadget and kernel seed snap info. In some cases we only want the components
+// of the essential snaps for a given mode.
 // TODO: make this method optionally return the system seed, since it might not
 // always be needed, and it is quite large.
-func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (*systemAndEssentialSnaps, error) {
+func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type, mode string) (*systemAndEssentialSnaps, error) {
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:2256]: should we be a bit more explicit and call mode  modeForComponents?

File: overlord/devicestate/handlers_install.go [hunk 8]
```diff
@@ -913,17 +925,75 @@ func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (*syste
...
+
+	return systemAndSnaps, mntPtForType, mntPtForComp, unmount, nil
+}
+
+func kModsInfo(systemAndSnaps *systemAndEssentialSnaps, kernMntPoint string, mntPtForComps map[string]string, isCore bool) (*install.KernelSnapInfo, []boot.BootableKModsComponents) {
+	kernInfo := systemAndSnaps.InfosByType[snap.TypeKernel]
+
+	// Find out kernel-modules components in the seed
+	compSeedInfos := systemAndSnaps.CompsByType[snap.TypeKernel]
+	bootKMods := make([]boot.BootableKModsComponents, 0, len(compSeedInfos))
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:968]: wouldn't make sense to add a field for this to KernelSnapInfo ? want we need it for when we deal with installing from snap-bootstrap or the API?

**Review by pedronis — COMMENTED** [review]
question, also what happens if the install is done by snap-bootstrap?

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, see my comment below. You are right that initramfs installation is missing, I've added a TODO there as this PR is big enough and I will do that as a follow-up.

**Comment by alfonsosanchezbeato** [comment]
I've added the new spread tests and rebased now (due to go version bump)

File: overlord/devicestate/handlers_install.go (hunk 8 above)
Comment (pedronis) [reply overlord/devicestate/handlers_install.go:968]: thanks, fair enough, please double check again then, maybe instead we need a wrapper struct that carries both?

Comment (alfonsosanchezbeato) [reply overlord/devicestate/handlers_install.go:968]: Exactly, I was thinking yesterday something like that

File: overlord/devicestate/handlers_install.go [hunk 9]
```diff
@@ -651,6 +664,7 @@ func (m *DeviceManager) doFactoryResetRunSystem(t *state.Task, _ *tomb.Tomb) err
 		UnpackedGadgetDir: gadgetDir,
 
 		RecoverySystemLabel: modeEnv.RecoverySystem,
+		KernelMods:          bootKMods,
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:667]: no test seems to fail in the package if this is not passed in

**Review by pedronis — COMMENTED** [review]
thank you, did another pass, mostly minor things but a question about tests

File: overlord/devicestate/handlers_install.go (hunk 9 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/handlers_install.go:662]: Test added

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, I've addressed your comments now

**Review by andrewphelpsj — APPROVED** [review]
Thanks, LGTM!

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**